lib/sqlalchemy/orm/attributes.py

```diff
@@ -270,6 +270,13 @@ class InstrumentedAttribute(QueryableAttribute):
     """
 
     def __set__(self, instance, value):
+        if self.prop in self.parent._readonly_props:
```

Comment:
the AttributeImpl should have a flag right up front called ".readonly", checks that flag.   the flag would be passed through when the impl is set up at https://github.com/sqlalchemy/sqlalchemy/blob/master/lib/sqlalchemy/orm/strategies.py#L100 .